### PR TITLE
Anemoi: single anemoi pemutation round gadget

### PR DESCRIPTION
**Pull request by vesselinux**
Addresses Issue https://github.com/clearmatics/libsnark/issues/96 .

### CODE REVIEW

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc [hunk 1]
```diff
@@ -321,32 +321,293 @@ void flystel_prime_field_gadget<ppT, parameters>::generate_r1cs_witness()
...
+        Z_left.push_back(X_left[i] + C[i]);
+        Z_right.push_back(X_right[i] + D[i]);
+    }
+
+    if (ncols > 1) {
+        M_matrix = anemoi_permutation_mds<ppT, ncols>(g);
+    } else { // ncols == 1
+        // the MDS matrix for a state with 1 column (L=1) is the same as
+        // for a state with 2 columns (L=2)
+        M_matrix = anemoi_permutation_mds<ppT, 2>(g);
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc:540]: Can't see where this is used.  Am I missing something?

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc [hunk 2]
```diff
@@ -321,32 +321,293 @@ void flystel_prime_field_gadget<ppT, parameters>::generate_r1cs_witness()
...
+        Z_left.clear();
+        Z_right.clear();
+        Z_left.push_back(Z[0]);
+        Z_right.push_back(Z[1]);
+    }
+
+    // apply layer of L Flystel S-boxes
+    for (size_t i = 0; i < ncols; i++) {
+        flystel_prime_field_gadget<ppT, parameters> H(
+            pb, Z_left[i], Z_right[i], Y_left[i], Y_right[i], "Flystel[i]");
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: annotation string is not formatted properly.  There is a macro FORMAT (or similar) for this.  It should be appended to the annotation string for the parent gadget.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc [hunk 3]
```diff
@@ -321,32 +321,293 @@ void flystel_prime_field_gadget<ppT, parameters>::generate_r1cs_witness()
...
+        Flystel.push_back(H);
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: Simpler to use `emplace_back`?

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/anemoi_outputs.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
+/** @file
+ *****************************************************************************
+ * @author     This file is part of libff, developed by Clearmatics Ltd
+ *             (originally developed by SCIPR Lab) and contributors
+ *             (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+#ifndef LIBSNARK_GADGETLIB1_GADGETS_HASHES_ANEMOI_TESTS_ANEMOI_OUTPUTS_CPP_
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/anemoi_outputs.cpp]: Should not need these guards in cpp files.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp [hunk 5]
```diff
@@ -224,18 +226,92 @@ void test_flystel_prime_field_gadget()
...
+    pb_variable_array<FieldT> Y_left;
+    pb_variable_array<FieldT> Y_right;
+
+    X_left.allocate(pb, NumStateColumns_L, "left inputs");
+    X_right.allocate(pb, NumStateColumns_L, "right inputs");
+
+    Y_left.allocate(pb, NumStateColumns_L, "left outputs");
+    Y_right.allocate(pb, NumStateColumns_L, "right outputs");
+
+    // WARNINIG! test with zero constants. TODO add the original
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp]: Should this be addressed?  Seems like we might not be sufficiently testing.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp [hunk 6]
```diff
@@ -224,18 +226,92 @@ void test_flystel_prime_field_gadget()
...
-    test_flystel_E_root_five_gadget<ppT, parameters>();
-    test_flystel_prime_field_gadget<ppT, parameters>();
+    test_flystel_E_root_five_gadget<ppT, parameters_debug>();
+    test_flystel_prime_field_gadget<ppT, parameters_debug>();
+    // Use the original parameters for the full permutation
+    using parameters = anemoi_parameters<ppT>;
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 1, parameters>();
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 2, parameters>();
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 3, parameters>();
+    // TODO code for L=4 is still WIP. The test values match, but a
```
Comment (dtebbs) [inline libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp]: Seems like this should be tracked down before merging this.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc (hunk 1 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc:540]: This clause will be activated for `ncols = NumStateColumns_L = 1` e.g. here https://github.com/clearmatics/libsnark/blob/anemoi-hash-r1cs-permutation-round/libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp#L308

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc (hunk 2 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: Fixed

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc (hunk 3 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: Fixed

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/anemoi_outputs.cpp (hunk 4 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/anemoi_outputs.cpp]: Removed

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp (hunk 5 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp]: Addressed.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp (hunk 6 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp]: Fixed.

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc (hunk 3 above)
Comment (dtebbs) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: You've just replaced `push_back` by `emplace_back`, which does not make anything simpler.  The point of `emplace_back` is that you don't have to create an intermediate `flystel_prime_field_gadget` object, you can pass the arguments to the constructor of the element directly into `emplace_back`.

**Review by dtebbs — CHANGES_REQUESTED** [review]
Looks good.  Just need to fix the use of `emplace_back`.

**Comment by vesselinux** [comment]
Moving discussion to new PR #101 after rebase on base branch `anemoi-hash-r1cs`

File: libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc (hunk 3 above)
Comment (vesselinux) [reply libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc]: Fixed. See #101.

**Pull request closed without merging**